### DISCUSSION TITLE
one-to-many self relations in Mongodb

### DISCUSSION THREAD

**Discussion by bekogeko** [prisma/prisma#17986] [Q&A]
first I want to give you some code then explain its use case and its benefits. Also this model is an outcome of #17746

```prisma
model Comment {
    id      String @id @default(auto()) @map("_id") @db.ObjectId

    // list of parent comments ordered by how deep they are
    // the first element is the parent comment
    // the last element is the root comment
    parentIds        String[]  @db.ObjectId
    parentComments   Comment[] @relation("ParentComments", fields: [parentIds ], references: [id])
}
```

this data structure benefits the deletion algorithm for some reasons. Let's say the comments in the post look like this and their ids are shown as #.n

```
- Root Comment #.1
- - Child Comment #.2
- - - Deep Child Comment #.4
- - Child Comment #.3
- Root Comment #.4
```
For example  the comment with id #.2 can be saved to db as 
```
id    2
parentIds { 1 }
```
And therefore if we want to delete the #.2 we can just say delete the comments which include #.2 in their parentIds. This is an optimization because we don't need to recursively delete them but this model not works and I don't know why. The model looks fine to me but schema says the otherwise.

I just need to record an array of ObjectIds in a model itself. Any idea on how to implement this data structure in the database?

**Comment by nurul3101** [comment]
Hello @bekogeko 👋 

Would this approach work for you?

```prisma
model Comment {
  id String @id @default(auto()) @map("_id") @db.ObjectId

  // list of parent comments ordered by how deep they are
  // the first element is the parent comment
  // the last element is the root comment
  parentIds      String[]  @db.ObjectId
  parentComments Comment[] @relation("Comments", fields: [parentIds], references: [id])
  childIds       String[]  @db.ObjectId
  childComments  Comment[] @relation("Comments", fields: [childIds], references: [id])
}
```

This is a bit modified and has a many-to-many self relations. The hierarchy of parent comments can be stored in `parentIds` array and the hierarchy of child comments can be stored in the `childIds` array.

**Reply by bekogeko** [reply]
Actually, I don't need child comments cant I ignore them?

**Reply by nurul3101** [reply]
If you don't need child comments, then you can remove the `childIds` and `childComments` field.

**Reply by bekogeko** [reply]
I can't actually because it gives an error 
```
Error validating field `parentComments` in model `Comment`: The relation field `parentComments` on model `Comment` is missing an opposite relation field on the model `Comment`. Either run `prisma format` or add it manually.
```

**Reply by nurul3101** [reply]
Are you removing both childIds and childComments?

**Reply by bekogeko** [reply]
No just childIds and childComments field

**Reply by bekogeko** [reply]
and still not working

**Comment by saibbyweb** [comment]
Stuck with this as well: Error validating field `XXX` in model `YYY`: The relation field `XXX` on model `YYY` is missing an opposite relation field on the model `YYY`. Either run `prisma format` or add it manually.

**Comment by saibbyweb** [comment]
```
model Comment {
  id String @id @default(auto()) @map("_id") @db.ObjectId
  parentIds      String[]  @db.ObjectId
  parentComments Comment[] @relation("Comments", fields: [parentIds], references: [id])
}
```
This schema results in: **Error validating field `parentComments` in model `Comment`: The relation field `parentComments` on model `Comment` is missing an opposite relation field on the model `Comment`. Either run `prisma format` or add it manually.**

then If I run `prisma format`, it becomes this:

```
model Comment {
  id             String    @id @default(auto()) @map("_id") @db.ObjectId
  parentIds      String[]  @db.ObjectId
  parentComments Comment[] @relation("Comments", fields: [parentIds], references: [id])
  Comment        Comment?  @relation(fields: [commentId], references: [id])
  commentId      String?   @db.ObjectId
}
```
which then result in two fields (**parentComments** & **Comment**) complaining about the missing opposite relation field?

The example in the official docs for 1:N self relations for MongoDB is this:

```
model User {
  id        String  @id @default(auto()) @map("_id") @db.ObjectId
  name      String?
  teacherId String? @db.ObjectId
  teacher   User?   @relation("TeacherStudents", fields: [teacherId], references: [id])
  students  User[]  @relation("TeacherStudents")
}
```

& it works. But what if I only need the *students* array field for the User model? Is there any provision for that, because this:

```
model User {
  id        String  @id @default(auto()) @map("_id") @db.ObjectId
  name      String?
  studentIds String[] @db.ObjectId
  students   User[]   @relation("students", fields: [studentIds], references: [id])
}
```
doesn't work!

**Reply by bekogeko** [reply]
doesnt work for me either

**Comment by daeteck** [comment]
Same issue here. Is there a solution to this? I see that the comments in this bug are getting old...

**Reply by bekogeko** [reply]
Still no answer

**Reply by janpio** [reply]
This is a discussion, not a bug. If there is a bug, someone should open a bug report and link to it here. Otherwise we can not take action on it.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
